pkg/logstructured/logstructured.go

```diff
@@ -294,28 +300,55 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 				}
 			}
 		}
+
+		logrus.Info("TTL events watch channel was closed")
 	}()
 
 	return result
 }
 
 func (l *LogStructured) ttl(ctx context.Context) {
 	// vary naive TTL support
-	mutex := &sync.Mutex{}
-	for event := range l.ttlEvents(ctx) {
-		go func(event *server.Event) {
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Duration(event.KV.Lease) * time.Second):
-			}
-			mutex.Lock()
-			if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-				logrus.Errorf("failed to delete expired key: %v", err)
-			}
-			mutex.Unlock()
-		}(event)
+	queue := workqueue.NewDelayingQueue()
+	go func() {
+		for l.handleTTLEvents(ctx, queue) {
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			queue.ShutDown()
+			return
+		default:
+		}
+
+		for event := range l.ttlEvents(ctx) {
+			queue.AddAfter(event, time.Duration(event.KV.Lease)*time.Second)
```

Comment:
I was thinking about this a bit more, and realized that we are missing a bit of nuance about how workqueues are normally used alongside an object cache.

We don't actually want to enqueue the full event here, just the event key. If we enqueue the event, then each update is queued for processing. We want to enqueue the key, so that each update to the TTL pushes out its time in the queue. By doing this, we can mostly avoid deleting old revisions entirely.

This does mean that we need to maintain a store parallel to the queue, that hold the most recently seen revision for each key. When the key does come up in the queue, we delete `key, revision` from the datastore, and remove the key from both the queue and the store.

[cache.NewStore()](https://pkg.go.dev/k8s.io/client-go/tools/cache#NewStore) is probably a good choice for this very simple use case.

Reply:
I understand what you mean. Your idea is actually to ignore intermediate versions of updated events, which would be more efficient. I think maybe we don't need to introduce cache.ThreadSafeStore directly in this simple case. Maybe go's native sync.Map can also solve this problem?

Firstly, we can't just enqueue the key into the queue, we need to put the struct{key, revision} into the queue. Because when an event is updated, we cannot directly update the enqueue time of the previous revision in the workqueue to the latest version. Therefore, after getting the event from the queue, we need to compare whether the key of the event exists in the map. If it exists, compare whether the revision is smaller than the latest revision in the map. If so, then we can ignore the event and avoid doing the delete operation.

What's your suggestion?

Reply:
> I think maybe we don't need to introduce cache.ThreadSafeStore directly in this simple case. Maybe go's native sync.Map can also solve this problem

I'm not sure, I'd have to take a closer look at it. I know that Kubernetes uses the cache library alongside workqueue, and I would expect there to be some specific reason for that.

> we cannot directly update the enqueue time of the previous version in the workqueue to the latest version as delaying queue only choose the minimum time.

Hmm, that is unfortunate. I was expecting workqueue to allow us to delay processing of an item by reenqueuing it with a higher delay. If this is not the case then some of its advantages are lost.

I am also realizing that the fact that we simultaneously list and watch will create problems. We need to either list then start a watch at the last revision seen by the watch (which is what Kubernetes does), or use a store type which understands key revisions. There is such a store available in the cache module, but it may be more correct to just list-then-watch.

Reply:
>> I think maybe we don't need to introduce cache.ThreadSafeStore directly in this simple case. Maybe go's native sync.Map can also solve this problem

> I'm not sure, I'd have to take a closer look at it. I know that Kubernetes uses the cache module alongside workqueue, and I would expect there to be some specific reason for that.

![image](https://github.com/k3s-io/kine/assets/136570416/1dd556c0-7d68-4a34-89ea-665ee2c74f41)

As simply shown in the figure above, workqueue and ThreadSafeStore are mainly used together for apiserver List-Watch processing logic in client-go. The main function of ThreadSafeStore is to store objects from list-watch and maintain complex indexes of objects and keys. Through the encapsulation of client-go, the desired object can be obtained indirectly from ThreadSafeStore according to the specified key or rule through Get/List and other interfaces in the custom controller.
In our scenario, we do not need to maintain the mapping from Object to key, and encapsulate the Get/List interface to query Object. In addition, WorkQueue and ThreadSafeStore are inherently independent, and there is no binding relationship when used. We only store the latest Revision corresponding to the Event's Key.

>> we cannot directly update the enqueue time of the previous version in the workqueue to the latest version as delaying queue only choose the minimum time.
> Hmm, that is unfortunate. I was expecting workqueue to allow us to delay processing of an item by reenqueuing it with a higher delay. If this is not the case then some of its advantages are lost.

The design of client-go's delaying work queue itself is to process the Object received from apiserver List-Watch. This design takes into account the final reconciliation consistency of the Object. When there is already an Item in the queue, the same Item will be re-enqueued. , he will compare the delay times of the two Items and choose the smallest time in order to process the Item faster. Reference code:
https://github.com/kubernetes/client-go/blob/33d14001dbdfd2f3eccfa907b8cc3738f514680e/util/workqueue/delaying_queue.go#L315C3-L318C4

> I am also realizing that the fact that we simultaneously list and watch will create problems due to old and new revisions being placed into the channel simultaneously. We need to either list then start a watch at the last revision seen by the watch (which is what Kubernetes does), or use a store type which understands key revisions. There is such a store available in the cache module, but it may be more correct to just list-then-watch.

I also noticed the problem of List and watch in the ttlEvents function of the current code. Although the best way is to adjust the order of List and Watch to ensure the order of elements placed in the delaying work queue. This problem only occurs when Kine is restarted. As a compromise method, a parallel store is introduced, the type is sync.Map, which stores the latest key and revision mapping from List and Watch. When the delaying work queue dequeues an element which contains {key, revision}, we need to compare whether the revision of the element is equal to the revision corresponding to the element in the store. If it is equal, then we actually call the Delete method to delete the element.

https://github.com/k3s-io/kine/blob/c5f2d892d70778fb9ad15208d4ca8617c3c794da/pkg/logstructured/logstructured.go#L269C1-L297C5

Reply:
Hi @brandond，
what do you think for above discussion ?

Reply:
>  When there is already an Item in the queue and the same Item be re-enqueued, it will compare the delay times of the two Items and choose the smallest time in order to process the Item faster.

It's unfortunate that the comparison here isn't a configurable property of the queue. It's probably not worth maintaining our own queue type just to change this:
https://github.com/kubernetes/client-go/blob/33d14001dbdfd2f3eccfa907b8cc3738f514680e/util/workqueue/delaying_queue.go#L312

If we store key+revision, then we are missing a lot of the benefits of the queue, as each key+revision will be a unique entry and still need to be processed. There will be no such thing as an item "already being in the queue" as every key+revision is guaranteed to be unique.

I think instead we should enqueue only the key, and maintain a map something like this:
```golang
type ttlKV struct {
  key string
  modRevision int64
  expires time.Time
}

keyMap := map[string]ttlKV
```

When we get an event, the TTL duration is added to the current time, to calculate `expires` for the map entry. When the key is dequeued, check if the expiration time has been reached. If so, delete the key+revision; if not re-enqueue the key.

> This problem only occurs when Kine is restarted.

I believe this is a problem any time the watch channel is closed and `ttlEvents` exits the loop, not just when kine is restarted. When this happens, we want to run `ttlEvents` again which will perform a new the list+watch, to ensure that we didn't miss anything. This issue is explicitly one of the things that you were hoping to fix, correct?

Reply:
Yes, I agree with your point of view. I will fix these problems later.

Reply:
Hi @brandond, 

I have updated, please help to have a review. Thanks.

Reply:
Hi @brandond,

I have updated this PR, please take a time to review. thanks.